web/popminer/dispatch.go

```diff
@@ -0,0 +1,438 @@
+	return Object{
+		"version":   version,
+		"gitCommit": gitCommit,
+	}.Value(), nil
+}
+
+func generateKey(_ js.Value, args []js.Value) (js.Value, error) {
+	log.Tracef("generatekey")
+	defer log.Tracef("generatekey exit")
+
+	var (
+		btcChainParams *btcchaincfg.Params
+		netNormalized  string
+	)
+	net := args[0].Get("network").String()
+	switch net {
+	case "devnet", "testnet3", "testnet":
+		btcChainParams = &btcchaincfg.TestNet3Params
+		netNormalized = "testnet3"
+	case "mainnet":
+		btcChainParams = &btcchaincfg.MainNetParams
+		netNormalized = "mainnet"
+	default:
+		return js.Null(), fmt.Errorf("invalid network: %v", net)
+	}
+
+	privKey, err := dcrsecpk256k1.GeneratePrivateKey()
+	if err != nil {
+		log.Errorf("failed to generate private key: %v", err)
+		return js.Null(), fmt.Errorf("generate secp256k1 private key: %w", err)
+	}
+	btcAddress, err := btcutil.NewAddressPubKey(
+		privKey.PubKey().SerializeCompressed(),
+		btcChainParams,
+	)
+	if err != nil {
+		log.Errorf("failed to generate btc address: %v", err)
+		return js.Null(), fmt.Errorf("create BTC address from public key: %v", err)
+	}
+
+	compressedPubKey := privKey.PubKey().SerializeCompressed()
+	ethereumAddress := ethereum.PublicKeyToAddress(compressedPubKey).String()
+
+	return Object{
+		"ethereumAddress": ethereumAddress,
+		"network":         netNormalized,
+		"privateKey":      hex.EncodeToString(privKey.Serialize()),
+		"publicKey":       hex.EncodeToString(compressedPubKey),
+		"publicKeyHash":   btcAddress.AddressPubKeyHash().String(),
+	}.Value(), nil
+}
+
+func startPoPMiner(_ js.Value, args []js.Value) (js.Value, error) {
+	log.Tracef("startPoPMiner")
+	defer log.Tracef("startPoPMiner exit")
+
+	pmMtx.Lock()
+	defer pmMtx.Unlock()
+	if pm != nil {
+		return js.Null(), errors.New("pop miner already started")
+	}
+
+	pm = new(Miner)
+	pm.ctx, pm.cancel = context.WithCancel(context.Background())
+
+	cfg := popm.NewDefaultConfig()
+	cfg.BTCPrivateKey = args[0].Get("privateKey").String()
+	cfg.StaticFee = uint(args[0].Get("staticFee").Int())
+
+	cfg.LogLevel = args[0].Get("logLevel").String()
+	if cfg.LogLevel == "" {
+		cfg.LogLevel = "popm=ERROR"
+	}
+	loggo.ConfigureLoggers(cfg.LogLevel)
+
+	network := args[0].Get("network").String()
+	netOpts, ok := networks[network]
+	if !ok {
+		return js.Null(), fmt.Errorf("unknown network: %s", network)
+	}
+	cfg.BFGWSURL = netOpts.bfgURL
+	cfg.BTCChainName = netOpts.btcChainName
+
+	var err error
+	pm.miner, err = popm.NewMiner(cfg)
+	if err != nil {
+		return js.Null(), fmt.Errorf("create PoP miner: %w", err)
+	}
+
+	// launch in background
+	pm.wg.Add(1)
+	go func() {
+		defer pm.wg.Done()
+		if err := pm.miner.Run(pm.ctx); !errors.Is(err, context.Canceled) {
+			// TODO: dispatch event on failure
+			pmMtx.Lock()
+			defer pmMtx.Unlock()
+			pm.err = err // Theoretically this can logic race unless we unset om
+		}
+	}()
+
+	return js.Undefined(), nil
+}
+
+func stopPopMiner(_ js.Value, _ []js.Value) (js.Value, error) {
```

Comment:
shouldn't this cancel the context and wg.Wait?

Reply:
I think it does? https://github.com/hemilabs/heminetwork/blob/joshua/web-popm/improve-go/web/popminer/dispatch.go#L306-L307